cmd/fleetctl/debug.go

```diff
@@ -316,3 +322,144 @@ func debugArchiveCommand() *cli.Command {
+			if err != nil {
+				return errors.Wrap(err, "Fail")
+			}
+			fmt.Fprintf(c.App.Writer, "Success: can resolve host %s.\n", baseURL.Hostname())
+			_ = ips
+
+			// 2. Attempt a raw TCP connection to host:port.
+			if err := dialHostPort(c.Context, timeoutPerCheck, baseURL.Host); err != nil {
+				return errors.Wrap(err, "Fail")
+			}
+			fmt.Fprintf(c.App.Writer, "Success: can dial server at %s.\n", baseURL.Host)
+
+			if cert := getFleetCertificate(c); cert != "" {
+				// 3. Is the certificate valid at all (x509.Certificate.ParseCertificate?)
+				// 4. Is the certificate valid for the hostname/IP address (x509.Certificate.VerifyHostname?)
+			}
+
+			// 5. Check that the server responds with expected responses (by
+			// making a POST to /api/v1/osquery/enroll with an invalid
+			// secret).
+			if err := checkAPIEndpoint(c.Context, timeoutPerCheck, fleet); err != nil {
+				return errors.Wrap(err, "Fail")
+			}
+			fmt.Fprintln(c.App.Writer, "Success: agent API endpoints are available.")
+
+			return nil
+		},
+	}
+}
+
+func resolveHostname(ctx context.Context, timeout time.Duration, host string) ([]net.IP, error) {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	var r net.Resolver
+	return r.LookupIP(ctx, "ip", host)
+}
+
+func dialHostPort(ctx context.Context, timeout time.Duration, addr string) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	var d net.Dialer
+	conn, err := d.DialContext(ctx, "tcp", addr)
+	if err == nil {
+		conn.Close()
+	}
+	return err
+}
+
+func checkAPIEndpoint(ctx context.Context, timeout time.Duration, client *service.Client) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	var enrollRes struct {
+		Error       string `json:"error"`
+		NodeInvalid bool   `json:"node_invalid"`
+	}
+	// make an enroll request with a deliberately invalid secret,
+	// to see if we get the expected error json payload.
+	res, err := client.DoContext(ctx, "POST",
+		"/api/v1/osquery/enroll", "", map[string]string{
+			"enroll_secret": "--invalid--",
+		})
+	if err != nil {
+		return errors.Wrap(err, "request failed")
+	}
+	defer res.Body.Close()
+
+	if err := json.NewDecoder(res.Body).Decode(&enrollRes); err != nil {
+		return errors.Wrap(err, "invalid JSON")
+	}
+	if res.StatusCode != http.StatusUnauthorized || enrollRes.Error == "" || !enrollRes.NodeInvalid {
+		return fmt.Errorf("unexpected %d response", res.StatusCode)
+	}
+	return nil
+}
+
+func checkFleetCert(ctx context.Context, timeout time.Duration, certPath, host string, ips []net.IP) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	//	b, err := ioutil.ReadFile(cert)
+	//	if err != nil {
+	//		return errors.Wrap(err, "invalid certificate path")
+	//	}
+	//	// TODO: flag is the PEM, not the DER
+	//	xcert, err := x509.ParseCertificate(b)
+	//	if err != nil {
+	//		return errors.Wrap(err, "parsing x509 cert")
+	//	}
+	//	_ = xcert
+	return nil
+}
```

Comment:
Check out https://github.com/fleetdm/fleet/blob/main/tools/osquery/fleet.crt for an example cert that we commonly use in development.

I also just remembered we have https://github.com/fleetdm/fleet/blob/main/orbit/pkg/certificate/certificate.go#L13-L64 which may be useful.

Reply:
Thanks, that's super helpful! A couple things about the code in `certificate.go`:

* Is this ok for fleet (fleetctl command) to use an orbit package? I remember reading somewhere that those two could be used independently, maybe we want to keep that boundary in the code dependencies too? Maybe it's fine now that they live in the same repo though.
* I'm not super familiar with the various `tls.Config` fields, but reading the docs (https://pkg.go.dev/crypto/tls#Config) it seems like in `ValidateConnection` it should use `RootCAs` instead of `ClientCAs` - RootCAs is for the clients, to validate server certificates, while ClientCAs is for the servers, to validate client-provided certificates IF the ClientAuth policy is not NoClientCert (which is the default).

I can address this as part of the PR (if there's indeed anything to change).

Reply:
As discussed in chat, this package can be moved to a top-level `pkg` directory. Fleet and Orbit should be able to function independently, but it's fine for them to share code dependencies.

In regards to `ClientCAs`, looks like you are right. https://github.com/fleetdm/fleet/blob/main/orbit/pkg/certificate/certificate.go#L50 is probably the line that has the effect we were actually looking for. Please go ahead and make the appropriate fixes.